## Data Collection and Pre Processing

In [1]:
# install and call required packages
from pandas_datareader import data, wb
import pandas as pd
import numpy as np
import seaborn as sns
import datetime as dt
import yfinance as yf
%matplotlib inline

Import banks historical stock data including start and end periods, OHLC data, stock dividen(optional), pre and post market data

In [2]:
# Create an DataFrame to store each bank stocks data
BAC = yf.download('BAC', start='2018-1-1', end='2023-12-31') # bank of america
C = yf.download('C', start='2018-1-1', end='2023-12-31') # citi group
GS = yf.download('GS', start='2018-1-1', end='2023-12-31') # goldman sachs
JPM = yf.download('JPM', start='2018-1-1', end='2023-12-31') #JP morgan
MS = yf.download('MS', start='2018-1-1', end='2023-12-31') # morgan stanley
WFC = yf.download('WFC', start='2018-1-1', end='2023-12-31') # well fargo

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


In [3]:
# Create a list of the ticker symbols (as strings)
tickers = ['BAC', 'C', 'GS', 'JPM', 'MS', 'WFC'] # group banks tickers in a listb

In [4]:
# concatenate the bank dataframes together to a single data frame called stocks. 
stocks = pd.concat([BAC, C, GS, JPM, MS, WFC],axis=1,keys=tickers)

In [5]:
# Set the column name levels
stocks.columns.names = ['Ticker','Stock Info']

### Clean and preprocess the data. Handle missing values, outliers, and ensure data types are correct.

In [6]:
# Check for missing values and Optionally, drop any remaining NaN values
bankStocks = stocks.dropna()
bankStocks.head()

Ticker            BAC                                                        \
Stock Info       Open       High        Low      Close  Adj Close    Volume   
Date                                                                          
2018-01-02  29.750000  29.900000  29.610001  29.900000  25.995726  57121600   
2018-01-03  29.900000  29.940001  29.690001  29.799999  25.908791  57865700   
2018-01-04  29.969999  30.440001  29.879999  30.190001  26.247862  76512500   
2018-01-05  30.370001  30.420000  30.049999  30.330000  26.369587  56445200   
2018-01-08  30.230000  30.270000  30.049999  30.120001  26.187008  42914800   

Ticker              C                                   ...         MS  \
Stock Info       Open       High        Low      Close  ...        Low   
Date                                                    ...              
2018-01-02  75.089996  75.180000  74.019997  74.360001  ...  52.169998   
2018-01-03  74.349998  74.739998  73.970001  74.589996  ...  51.919998   
2018-01-04  75.010002  76.110001  74.660004  75.510002  ...  52.509998   
2018-01-05  75.709999  75.769997  74.959999  75.410004  ...  52.740002   
2018-01-08  75.169998  75.239998  74.330002  74.529999  ...  52.689999   

Ticker                                           WFC                        \
Stock Info      Close  Adj Close   Volume       Open       High        Low   
Date                                                                         
2018-01-02  52.189999  43.752396  6644600  61.040001  61.360001  60.700001   
2018-01-03  52.330002  43.869762  6955500  61.220001  61.770000  61.099998   
2018-01-04  53.130001  44.540421  8678800  61.980000  63.049999  61.910000   
2018-01-05  53.110001  44.523651  6476000  62.759998  62.810001  62.090000   
2018-01-08  52.889999  44.339211  6681400  62.660000  62.730000  61.939999   

Ticker                                      
Stock Info      Close  Adj Close    Volume  
Date                                        
2018-01-02  61.090000  51.275238  13819300  
2018-01-03  61.560001  51.669735  14203700  
2018-01-04  62.330002  52.316025  18740500  
2018-01-05  62.750000  52.668533  14217900  
2018-01-08  62.040001  52.072609  15569400  

[5 rows x 36 columns]

### EDA
Let's explore the data a bit! Before continuing...
Extract the 'Close' prices from the 'Stock Info' level of the DataFrame and then calculate the maximum value for each bank's stock. 

In [7]:
# Maximum Closing price for each bank's stock throughout the time period
bankStocks.xs(key='Close', axis=1, level='Stock Info').max()

Ticker
BAC     49.380001
C       81.910004
GS     423.850006
JPM    171.779999
MS     108.730003
WFC     65.930000
dtype: float64

The highest closing price was Goldman Sach (GS) closing at USD 423.85 dollars, while the next highest is JP Morgan (JPM) closing at USD 171.78

In [8]:
# Minimum Closing price for each bank's stock throughout the time period
bankStocks.xs(key='Close', axis=1, level='Stock Info').min()

Ticker
BAC     18.080000
C       35.389999
GS     134.970001
JPM     79.029999
MS      27.809999
WFC     21.139999
dtype: float64

The lowest closing price for the period was Bank of America (BAC) with stock closing at USD 18.08, while the lowest for Citigroup (C) was USD 35.39, 